First we have to import libraries such as the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

Dataset provided in the assignment are:

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate

To read the dataset into a <em>pandas</em> dataframe:

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So, the data suggests that the first sample has 540 cubic metres of cement, 0 cubic metres of blast furnace slag, 0 cubic metre of fly ash, 162 cubic meter of water, 2.5 cubic metres of superplaticizer, 1040 cubic metres of coarse aggregate and, 676 cubic metres of fine aggregate. 
This concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

<h4>So to check the number of data points:</h4>

In [3]:
concrete_data.shape

(1030, 9)

So, there are total of 1030 samples to train our model on. So, we should be very careful not to overfit the training data.

We can use <i><u>concrete_data.describe()</u></i> to check the dataset for any missing values.

In [4]:
concrete_data.describe()
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

So we can colclude that this dataset is very clean and thus it can be used safely to build our model.

<h4>Split data into predictors and target</h4>

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [5]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # to get all columns except Strength
target = concrete_data['Strength'] # to get Strength column

To make a sanity check  forthe data again for sake of the predictors and the target dataframes.

In [6]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

So, it looks clean again.

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [8]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [9]:
number_cols = predictors_norm.shape[1] # number of predictors
number_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Importing Keras

#### Now we have to import the Keras library for the actual work

In [10]:
import keras

To use Keras, we first have to install the TensorFlow backend, which would be used to install the Keras library.

Now, to import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [11]:
from keras.models import Sequential
from keras.layers import Dense

In [12]:
def regression_model():
    # creating the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(number_cols,)))
    model.add(Dense(1))
    
    # compiling the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The regression_model function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [13]:
from sklearn.model_selection import train_test_split

Now, let's split the data into a training and test sets by holding 30% of the data for testing

In [14]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)


### Training and Testing the Network

Call the function now to create our model.

In [15]:
# building the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [16]:
# fiting the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 0s - loss: 1593.9716 - 398ms/epoch - 17ms/step
Epoch 2/50
23/23 - 0s - loss: 1575.6439 - 16ms/epoch - 694us/step
Epoch 3/50
23/23 - 0s - loss: 1557.7467 - 19ms/epoch - 824us/step
Epoch 4/50
23/23 - 0s - loss: 1539.5619 - 24ms/epoch - 1ms/step
Epoch 5/50
23/23 - 0s - loss: 1521.3284 - 29ms/epoch - 1ms/step
Epoch 6/50
23/23 - 0s - loss: 1503.2220 - 20ms/epoch - 862us/step
Epoch 7/50
23/23 - 0s - loss: 1484.4989 - 27ms/epoch - 1ms/step
Epoch 8/50
23/23 - 0s - loss: 1465.1234 - 20ms/epoch - 889us/step
Epoch 9/50
23/23 - 0s - loss: 1445.5850 - 28ms/epoch - 1ms/step
Epoch 10/50
23/23 - 0s - loss: 1425.1364 - 20ms/epoch - 867us/step
Epoch 11/50
23/23 - 0s - loss: 1403.9861 - 20ms/epoch - 867us/step
Epoch 12/50
23/23 - 0s - loss: 1382.2852 - 19ms/epoch - 822us/step
Epoch 13/50
23/23 - 0s - loss: 1359.7911 - 33ms/epoch - 1ms/step
Epoch 14/50
23/23 - 0s - loss: 1336.7854 - 22ms/epoch - 954us/step
Epoch 15/50
23/23 - 0s - loss: 1312.9774 - 23ms/epoch - 997us/step
Epoch 16/50
23

Next we need to evaluate the model on the test data.

In [17]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 826us/step


411.5313415527344

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [18]:
from sklearn.metrics import mean_squared_error

In [19]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

411.5313451982344 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [20]:
total_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(
        predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " + str(total_mean_squared_errors) +
      " mean squared errors with normalized data. Total number of epochs for each training is: " + str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 114.78933715820312
10/10 [==============================] - 0s 772us/step
MSE 2: 82.453369140625
10/10 [==============================] - 0s 717us/step
MSE 3: 45.580142974853516
10/10 [==============================] - 0s 776us/step
MSE 4: 41.74267578125
10/10 [==============================] - 0s 0s/step
MSE 5: 40.5206298828125
10/10 [==============================] - 0s 841us/step
MSE 6: 43.09029006958008
10/10 [==============================] - 0s 2ms/step
MSE 7: 41.769004821777344
10/10 [==============================] - 0s 2ms/step
MSE 8: 32.778472900390625
10/10 [==============================] - 0s 0s/step
MSE 9: 34.11531066894531
10/10 [==============================] - 0s 776us/step
MSE 10: 34.24369430541992
10/10 [==============================] - 0s 940us/step
MSE 11: 34.17572021484375
10/10 [==============================] - 0s 662us/step
MSE 12: 29.907691955566406
10/10 [==============================] - 0s 776us/step
MSE 13: 38.792686462402344
10/10 [==============